common:
    image_left_topic: "/cam0/image_raw"
    image_right_topic: "/cam1/image_raw"
    imu_topic: "/imu0"
    
state_parameter:
    use_fej: true                        # if first-estimate Jacobians should be used (enable for good consistency)
    calib_cam_extrinsics: false          # if the transform between camera and IMU should be optimized R_ItoC, p_CinI
    calib_cam_intrinsics: true           # if camera intrinsics should be optimized (focal, center, distortion)
    calib_cam_timeoffset: true           # if timeoffset between camera and IMU should be optimized
    calib_imu_intrinsics: false          # if imu intrinsics should be calibrated (rotation and skew-scale matrix)
    calib_imu_g_sensitivity: false       # if gyroscope gravity sensitivity (Tg) should be calibrated
    imu_intrinsics_model: "kalibr"       # "calibrated" (same as "kalibr"), "kalibr", "rpng"
    max_clones: 11                       # how many clones in the sliding window raw: 11
    max_slam: 100                        # number of features in our state vector raw: 50 200
    max_slam_in_update: 100              # update can be split into sequential updates of batches, how many in a batch raw: 25 200
    max_msckf_in_update: 100             # how many MSCKF features to use in the update raw: 50 200
    
initializer_parameter:
    init_window_time: 2.0                # how many seconds to collect initialization information 10.0 for MH_01, MH_02, MH_04, 2.0 for others
    init_imu_thresh: 0.5                 # threshold for variance of the accelerometer to detect a "jerk" in motion
    init_max_disparity: 1.5              # max disparity to consider the platform stationary (dependent on resolution)
    init_max_features: 50                # how many features to track during initialization (saves on computation)
    init_dyn_use: true                   # if dynamic initialization should be used
    init_dyn_mle_opt_calib: false        # if we should optimize calibration during intialization (not recommended)
    init_dyn_mle_max_iter: 50            # how many iterations the MLE refinement should use (zero to skip the MLE)
    init_dyn_mle_max_time: 0.05          # how many seconds the MLE should be completed in
    init_dyn_mle_max_threads: 6          # how many threads the MLE should use
    init_dyn_num_pose: 6                 # number of poses to use within our window time (evenly spaced)
    init_dyn_min_deg: 5.0                # orientation change needed to try to init
    init_dyn_inflation_ori: 10           # what to inflate the recovered q_GtoI covariance by
    init_dyn_inflation_vel: 100          # what to inflate the recovered v_IinG covariance by
    init_dyn_inflation_bg: 10            # what to inflate the recovered bias_g covariance by
    init_dyn_inflation_ba: 100           # what to inflate the recovered bias_a covariance by
    init_dyn_min_rec_cond: 1e-20         # reciprocal condition number thresh for info inversion
    init_dyn_bias_g: [ 0.0, 0.0, 0.0 ]   # initial gyroscope bias guess
    init_dyn_bias_a: [ 0.0, 0.0, 0.0 ]   # initial accelerometer bias guess
    gravity_mag: 9.80766                 # magnitude of gravity in this location
    downsample_cameras: false               # will downsample image in half if true

imu_parameter:
    accelerometer_noise_density: 5.8860e-03     # [m/s^2/sqrt(Hz)]   ( accel "white noise" )
    accelerometer_random_walk: 1.0000e-04       # [m/s^3/sqrt(Hz)]   ( accel bias diffusion )
    gyroscope_noise_density: 1.7453e-04         # [rad/s/sqrt(Hz)]   ( gyro "white noise" )
    gyroscope_random_walk: 1.0000e-05           # [rad/s^2/sqrt(Hz)] ( gyro bias diffusion )
    sigma_pix: 1.0
    Tw: [ 1.0, 0.0, 0.0, 
          0.0, 1.0, 0.0, 
          0.0, 0.0, 1.0 ]
    Ta: [ 1.0, 0.0, 0.0, 
          0.0, 1.0, 0.0, 
          0.0, 0.0, 1.0 ]
    R_acc_imu: [ 1.0, 0.0, 0.0, 
                 0.0, 1.0, 0.0, 
                 0.0, 0.0, 1.0 ]
    R_gyr_imu: [ 1.0, 0.0, 0.0, 
                 0.0, 1.0, 0.0, 
                 0.0, 0.0, 1.0 ]
    Tg: [ 0.0, 0.0, 0.0, 
          0.0, 0.0, 0.0, 
          0.0, 0.0, 0.0 ]
    
camera_parameter:
    timeshift_cam_imu_left: 0.0
    timeshift_cam_imu_right: 0.0
    distortion_model_left: radtan
    distortion_model_right: radtan
    intrinsics_left: [816.903790, 811.568038, 608.507263, 263.475998]   # fx, fy, cx, cy
    intrinsics_right: [813.782055, 808.521656, 613.864195, 249.410493]   # fx, fy, cx, cy
    distortion_coeffs_left: [-0.0561430, 0.139526, -0.00121559, -0.000972814]
    distortion_coeffs_right: [-0.0549220, 0.142437, 0.0000754123, -0.000675605]
    resolution_left: [1280, 560]
    resolution_right: [1280, 560]
    T_imu_cam_left: [-0.00681,-0.01532,0.99987,1.71239,  
                     -0.99998,0.00033,-0.00680,0.24740, 
                     -0.00023,-0.99988,-0.01532,-0.11589, 
                     0.0, 0.0, 0.0, 1.0]
    T_imu_cam_right: [-0.01036,-0.01075,0.99990,1.70544, 
                      -0.99994,-0.00178,-0.01038,-0.22770, 
                      0.00189,-0.99994,-0.01073,-0.11611, 
                      0.0, 0.0, 0.0, 1.0]
                      
odometry_parameter:
    use_mask: false
    use_huber: false
    use_keyframe: true
    dt_slam_delay: 1
    up_msckf_sigma_px: 1.5
    up_msckf_chi2_multipler: 1
    up_slam_sigma_px: 1.5
    up_slam_chi2_multipler: 1
    downsample_cameras: false
    num_pts: 500  # raw: 200 1000
    fast_threshold: 20
    patch_size_x: 5
    patch_size_y: 5
    min_px_dist: 20 # 20
    histogram_method: "histogram"
    track_frequency: 31.0
    
feature_parameter:
    refine_features: true
    max_runs: 5
    init_lamda: 1e-3
    max_lamda: 1e10
    min_dx: 1e-6
    min_dcost: 1e-6
    lam_mult: 10.0
    min_dist: 0.25
    max_dist: 150.0
    max_baseline: 200
    max_cond_number: 20000.0
    keyframe_parallax: 20

voxel_parameter:
    voxel_size: 1.0             # raw: 0.1
    use_all_points: false
    max_num_points_in_voxel: 5
    min_distance_points: 0.0    # raw: 0.03
    nb_voxels_visited: 2
